File: src/server/template_server.cpp
#include "behavior_trees/rosaction.h"
#include "behavior_trees/robot_config.h"

// simple template to implement actions and conditions for the
// platform being used, the execute callback is going to be
// executed with frenquency: EXECUTION_FREQUENCY

// class definition overriding the functions executeCB and resetCB
class ActionName : ROSAction
{
public:
	ActionName(std::string name) :
		ROSAction(name),
		time_to_complete_((ros::Duration) 0) // used for this example
		{}

	// the action succeeds automatically after 5 seconds
	int executeCB(ros::Duration dt)
		{
			std::cout << "**ActionName -%- Executing Main Task, elapsed_time: "
			          << dt.toSec() << std::endl;
			std::cout << "**ActionName -%- elapsed_time: "
			          << time_to_complete_.toSec() << std::endl;

			time_to_complete_ += dt;

			if (time_to_complete_.toSec() < 5)
			{
				set_feedback(RUNNING);
				// feedback_.FEEDBACK_ = RUNNING;
				// as_.publishFeedback(feedback_);
				return 0;		// 'allows' this thread to continue running
			}
			else if (time_to_complete_.toSec() >= 5)
			{
				set_feedback(SUCCESS);
				// set_feedback(FAILURE);
				// feedback_.FEEDBACK_ = FAILURE;
				// as_.publishFeedback(feedback_);
				// result_.RESULT_ = FAILURE;
				// as_.setSucceeded(result_);
				// stop();			// stop allows new threads to be created
				return 1;		// 'forces' this thread to finish securely
			}
			return 0;
		}

	void resetCB()
		{
			time_to_complete_ = (ros::Duration) 0;
		}

	// member variables
	ros::Duration time_to_complete_;
};

int main(int argc, char** argv)
{
	std::cout << "Hello, world!" << std::endl;
	// specify which options are available as cmd line arguments
	setupCmdLineReader();
	// read agent id from command line parameters (--agent=mario)
	std::string agent = readAgentFromCmdLine(argc, argv);
	// initialize the behavior tree server node
	ros::init(argc, argv, std::string("action_name") + "_" + agent); // name used for bt.txt
	ActionName server(ros::this_node::getName());
	ros::spin();
	return 0;
}

File: src/server/template_server.py
"""Template server file.

Provides an example of a python behavior trees server
"""
# ! /usr/bin/env python

import rospy
import roslib
from rospyaction import ROSPYAction, RUNNING, FAILURE, SUCCESS
roslib.load_manifest('behavior_trees')


class ActionName(ROSPYAction):
    """Class definition implementing the functions execute_cb and reset_cb."""

    def __init__(self, name):
        """ActionName constructor."""
        ROSPYAction.__init__(self, name)
        self._time_to_complete = rospy.Duration(0)

    def execute_cb(self, dt):
        """Run the action logic."""
        print "Executing Main Task, elapsed_time: %f" % dt.to_sec()

        self._time_to_complete = self._time_to_complete + dt

        if self._time_to_complete.to_sec() < 1.5:
            self.set_feedback(RUNNING)
            return 0  # Returning 0 keeps the action alive

        else:
            self.set_feedback(SUCCESS)
            return 1

    def reset_cb(self):
        """Reset the action time."""
        print "Executing the reset callback"
        self._time_to_complete = rospy.Duration(0)

if __name__ == '__main__':
    rospy.init_node('action_name_default')
    ActionName(rospy.get_name())
    rospy.loginfo('starting node %s' % rospy.get_name())
    rospy.spin()
